File: luogu/P1551亲戚.py
#!~/AppData/Local/Programs/Python/Python38-32/python.exe
# Disjoint Set
fa = []
rank = []

def init():
    for i in range(5010):
        fa.append(i)
        rank.append(1)

def find(x):
    if fa[x] == x:
        return x
    else:
        return find(fa[x])

def union(x, y):
    i = find(x)
    j = find(y)

    if rank[i] <= rank[j]:
        fa[i] = j
    else:
        fa[j] = i
        
    if rank[i] == rank[j] and i != j:
        rank[j] = rank[j] + 1

n, m, p = map(int, input().split())

init()

for i in range(m):
    x, y = map(int, input().split())
    union(x, y)

for i in range(p):
    x, y = map(int, input().split())
    if find(x) == find(y):
        print('Yes')
    else:
        print('No')

File: luogu/P1551亲戚.cc
/*
 *	Disjoint Set
 */
#include <iostream>
#include <algorithm>
#include <cstring>
#include <string>

using namespace std;

const int N = 5010;

const string _y = "Yes";
const string _n = "No";

int fa[N];
int _rank[N];

inline void init() {
	for(int i = 0; i < N; i++) {
		fa[i] = i;
		_rank[i] = 1;
	}
}

inline int find(int x) {
	if(fa[x] == x)
		return x;
	else
		return find(fa[x]);
}

inline void merge(int i, int j) {
	int x = find(i), y = find(j);
	if(_rank[x] <= _rank[y])
		fa[x] = y;
	else
		fa[y] = x;
	if(_rank[x] == _rank[y] && x != y)
		_rank[y]++;
}

int main() {
	int n, m, p, x, y;
	cin >> n >> m >> p;

	//Initalize the disjoint set
	init();

	for(int i = 0; i < m; i++) {
		cin >> x >> y;
		merge(x, y);
	}

	for(int i = 0; i < p; i++) {
		cin >> x >> y;
		if(find(x) != find(y))
			cout << _n << endl;
		else
			cout << _y << endl;
	}

	return 0;
}
